text/0000-javascript-module-api.md

```diff
@@ -0,0 +1,1058 @@
+`Ember`.)
+
+Using Ember via a monolithic package or global object is not ideal for several
+reasons:
+
+* It's overwhelming for learners. There's a giant list of classes and functions
+  with no hints about how they're related. The API documentation reflects this.
+* Experienced developers who don't want all of Ember's features feel like
+  they're adding unnecessary and inescapable bloat to their application.
+* The `Ember` object must be built at boot time, requiring that we ship the
+  entire framework to the browser. This has two major costs:
+  1. Increased download time, particularly noticeable on slower connections.
+  2. Increased parsing/evaluation cost, which still must be paid even when
+     assets are cached. On some browsers/devices, this can far exceed the cost of the
+     download itself.
+* Running multiple versions of Ember on the same page is error-prone.
+
+Defining a public API for importing parts of Ember via JavaScript modules helps
+us lay the groundwork for solving all of these problems.
+
+#### Multiple Apps on a Page
+
+Because modules do not export a global, there is no risk of a conflict running
+two or more Ember apps on the same page; each app can provide its own version
+without the risk of clobbering another.
+
+#### Reducing Load Time
+
+Modules help us eliminate unneeded code. The module syntax is _statically
+analyzable_, meaning that a tool like Ember CLI can analyze an application's
+source code and reliably determine which files, in both the framework and the
+application, are actually needed. Anything that's not needed is omitted from the
+final build.
+
+This allows us to provide the file size benefits of a "small modules" approach
+to building web applications while retaining the productivity benefits of a
+complete, opinionated framework.
+
+For example, if your application never used the `Ember.computed.union` computed
+property helper, Ember could detect this and remove its code when you build your
+application. This technique for slimming down the payload automatically is often
+referred to as _tree shaking_ or _dead code elimination_.
+
+Building the module graph doesn't just mean we get a list of files used by the application—
+we also know which files are used _route-by-route_.
+
+We can use this knowledge to optimize boot time even more, by prioritizing
+sending only the JavaScript needed for the requested route, rather than the
+entire application.
+
+For example, if the user requests the URL
+`https://app.example.com/articles/123`, the server could first send the code for
+`ArticlesRoute`, the `Article` model, the `articles` template, and any
+components and framework code used in the route. Only after the route is
+rendered would we start to send the remainder of the application and framework
+code in the background.
+
+#### Guiding Learners
+
+We can group framework classes and utilities by functionality, making it clear
```

Comment:
This is huge, framework discovery through intuition is a huge part of how devs learn. This also has the side benefit of making the documentation much easier to find.